server/neptune/gateway/event/root_config_builder.go

```diff
@@ -0,0 +1,86 @@
+package event
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/core/config"
+	"github.com/runatlantis/atlantis/server/core/config/valid"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/logging"
+)
+
+// repoFetcher manages a cloned repo's workspace on disk for running commands.
+type repoFetcher interface {
+	Fetch(ctx context.Context, baseRepo models.Repo, sha string) (string, func(ctx context.Context, filePath string), error)
+}
+
+// hooksRunner runs preworkflow hooks for a given repository/commit
+type hooksRunner interface {
+	Run(repo models.Repo, repoDir string) error
+}
+
+// fileFetcher handles being able to identify and fetch the changed files per individual commit
+type fileFetcher interface {
+	GetModifiedFilesFromCommit(ctx context.Context, repo models.Repo, sha string, installationToken int64) ([]string, error)
+}
+
+// rootFinder determines which roots were modified in a given event.
+type rootFinder interface {
+	// DetermineRoots returns the list of roots that were modified
+	// based on modifiedFiles and the repo's config.
+	DetermineRoots(modifiedFiles []string, config valid.RepoCfg) ([]valid.Project, error)
+}
+
+// parserValidator config builds repo specific configurations
+type parserValidator interface {
+	ParseRepoCfg(absRepoDir string, repoID string) (valid.RepoCfg, error)
+}
+
+type RootConfigBuilder struct {
+	RepoFetcher     repoFetcher
+	HooksRunner     hooksRunner
+	ParserValidator parserValidator
+	RootFinder      rootFinder
+	FileFetcher     fileFetcher
+	GlobalCfg       valid.GlobalCfg
+	Logger          logging.Logger
+}
+
+func (b *RootConfigBuilder) Build(ctx context.Context, event Push) ([]*valid.MergedProjectCfg, error) {
+	// Generate a new filepath location and clone repo into it
+	repoDir, deleteFn, err := b.RepoFetcher.Fetch(ctx, event.Repo, event.Sha)
```

Comment:
nit: I'm not personally a huge fan of having the type in the name unless there's no option (ie. duplicate vars with different types). With context for example, you get 

```go
ctx, cancel := context.WithCancel(...)
```

and cancel is the function.

I'd suggest using either `delete()` or `cleanup()` for this one if possible